lib/spack/spack/cmd/mirror.py

```diff
@@ -97,6 +98,19 @@ def setup_parser(subparser):
         default=spack.cmd.default_list_scope(),
         help="configuration scope to read from")
 
+    # Binary
+    upload_parser = sp.add_parser('binary', help=mirror_binary.__doc__)
+    upload_parser.add_argument('-d', '--directory', default=None,
+                               help="directory containing binaries for mirroring")
+    upload_parser.add_argument('-f', '--file', default=None,
+                               help="a file to upload to the mirror")
+    upload_parser.add_argument('-k', '--key', default=None,
+                               help="the s3 key")
+    upload_parser.add_argument('-b', '--bucket', default=None,
```

Comment:
The other mirror commands operate on mirror *names* (supplied via `spack mirror add`), not on raw urls. So I think the user should first `spack mirror add <name> <s3 url>`, then run `spack mirror put <specs>` to upload specs to that mirror. Ideally that would work with local mirrors in the file system and with s3 buckets.